mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,507 @@
+# Here, we purposefully stay away from shorthand notation in both and use
+# explicit terms (like 'azimuth' and 'polar') to avoid confusion.
+
+
+from __future__ import division
+from os import path as op
+import numpy as np
+from scipy.special import lpmv
+from scipy.linalg import pinv
+from scipy.misc import factorial as fact
+from ..utils import logger
+from .. import pick_types, pick_info
+from ..io.constants import FIFF
+from ..forward._compute_forward import _concatenate_coils
+from ..forward._make_forward import _read_coil_defs, _create_coils
+
+
+def maxwell_filter(raw, origin=(0, 0, 40), int_order=8, ext_order=3):
+    """Apply Maxwell filter to data using spherical harmonics.
+
+    Parameters
+    ----------
+    raw : instance of mne.io.Raw
+        Data to be filtered
+    origin : array-like, shape (3,)
+        Origin of internal and external multipolar moment space in head coords
+        and in millimeters
+    int_order : int
+        Order of internal component of spherical expansion
+    ext_order : int
+        Order of external component of spherical expansion
+
+    Returns
+    -------
+    raw_sss : instance of mne.io.Raw
+        The raw data with Maxwell filtering applied
+    """
+
+    # TODO: Exclude 'bads' in multipolar moment calc, add back during
+    # reconstruction
+    assert (raw.info['bads'] == [], 'Maxwell filter does not yet handle bad '
+            'channels.')
+
+    # Read coil definitions from file
+    coil_def_file = op.join(op.split(__file__)[0], '..', 'data',
+                            'coil_def_Elekta.dat')
+    all_coils, meg_info = _make_coils(raw.info, accurate=True,
+                                      coil_def=coil_def_file)
+
+    # Create coil list and pick MEG channels
+    picks = [raw.info['ch_names'].index(coil['chname'])
+             for coil in all_coils]
+    coils = [all_coils[ci] for ci in picks]
+    raw.pick_channels(coil['chname'] for coil in coils)
+
+    data, times = raw[picks, :]
+
+    # Magnetometers (with coil_class == 1.0) must be scaled by 100 to improve
+    # numerical stability as they have different scales than gradiometers
+    coil_scale = np.ones(len(picks))
+    coil_scale[np.array([coil['coil_class'] == 1.0 for coil in coils])] = 100.
+
+    # Compute multipolar moment bases
+    origin = np.array(origin) / 1000.  # Convert scale from mm to m
+    S_in, S_out = _sss_basis(origin, coils, int_order, ext_order)
+    S_tot = np.c_[S_in, S_out]
+
+    # Pseudo-inverse of total multipolar moment basis set (Part of Eq. 37)
+    pS_tot = pinv(S_tot, cond=1e-15)
+    # Compute multipolar moments of (magnetometer scaled) data (Eq. 37)
+    mm = np.dot(pS_tot, data * coil_scale[:, np.newaxis])
+    # Reconstruct data from internal space (Eq. 38)
+    recon = np.dot(S_in, mm[:S_in.shape[1], :])
+
+    # Return reconstructed raw file object
+    raw_sss = raw.copy()
+    raw_sss[:, :] = recon / coil_scale[:, np.newaxis]
```

Comment:
And I'm actually a bit surprised that this works... I would have assumed it needed to be `raw_sss._data[:, :] = ...`...?